File: services/frontend/src/lib/booking.ts
import api from './api';
import {
  Zone,
  Place,
  Slot,
  Booking,
  BookingCreate,
  BookingCreateByTime,
  BookingCancel,
  ZoneCreate,
  ZoneUpdate,
  ZoneCloseRequest,
  ZoneStatistics,
} from '@/types';

export const bookingService = {
  // Получить список зон
  async getZones(): Promise<Zone[]> {
    const response = await api.get('/bookings/zones');
    return response.data;
  },

  // Получить места в зоне
  async getPlacesByZone(zoneId: number): Promise<Place[]> {
    const response = await api.get(`/bookings/zones/${zoneId}/places`);
    return response.data;
  },

  // Получить доступные слоты для места на дату
  async getSlots(placeId: number, date: string): Promise<Slot[]> {
    const response = await api.get(`/bookings/places/${placeId}/slots`, {
      params: { date },
    });
    return response.data;
  },

  // Создать бронирование
  async createBooking(data: BookingCreate): Promise<Booking> {
    const response = await api.post('/bookings/', data);
    return response.data;
  },

  // Создать бронирование по времени
  async createBookingByTime(data: BookingCreateByTime): Promise<Booking> {
    const response = await api.post('/bookings/by-time', data);
    return response.data;
  },

  // Отменить бронирование
  async cancelBooking(data: BookingCancel): Promise<Booking> {
    const response = await api.post('/bookings/cancel', data);
    return response.data;
  },

  // Получить историю бронирований
  async getBookingHistory(filters?: {
    status?: string;
    zone_id?: number;
    date_from?: string;
    date_to?: string;
  }): Promise<Booking[]> {
    const response = await api.get('/bookings/history', {
      params: filters,
    });
    return response.data;
  },

  // Продлить бронирование
  async extendBooking(bookingId: number, extendHours: number = 1, extendMinutes: number = 0): Promise<Booking> {
    const response = await api.post(`/bookings/${bookingId}/extend`, {
      extend_hours: extendHours,
      extend_minutes: extendMinutes,
    });
    return response.data;
  },
};

// Admin функции
export const adminService = {
  // Получить все зоны (включая закрытые) для админа
  async getAdminZones(): Promise<Zone[]> {
    const response = await api.get('/admin/zones');
    return response.data;
  },

  // Создать зону
  async createZone(data: ZoneCreate): Promise<Zone> {
    const response = await api.post('/admin/zones', data); // <-- исправлено (убран /bookings)
    return response.data;
  },

  // Обновить зону
  async updateZone(zoneId: number, data: ZoneUpdate): Promise<Zone> {
    const response = await api.patch(`/admin/zones/${zoneId}`, data); // <-- исправлено
    return response.data;
  },

  // Удалить зону
  async deleteZone(zoneId: number): Promise<void> {
    await api.delete(`/admin/zones/${zoneId}`); // <-- исправлено
  },

  // Закрыть зону на обслуживание
  async closeZone(
    zoneId: number,
    data: ZoneCloseRequest
  ): Promise<Booking[]> {
    const response = await api.post(
      `/admin/zones/${zoneId}/close`, // <-- исправлено
      data
    );
    return response.data;
  },

  // Получить статистику по всем зонам
  async getZonesStatistics(): Promise<ZoneStatistics[]> {
    const response = await api.get('/admin/zones/statistics');
    return response.data;
  },
};

// Notification функции
export const notificationService = {
  // Массовая рассылка email всем пользователям (только для админа)
  async sendBulkEmail(subject: string, text: string): Promise<any> {
    const response = await api.post('/notifications/bulk', {
      subject,
      text,
    });
    return response.data;
  },

  // Получить уведомления пользователя
  async getUserNotifications(userId: number): Promise<any[]> {
    const response = await api.get(`/notifications/user/${userId}`);
    return response.data;
  },
};

File: services/api-gateway/routes/booking.py
from fastapi import APIRouter, Request, Depends, Response
import requests
from config import BOOKING_SERVICE_URL
from auth import get_current_user

router = APIRouter()

@router.get("/zones")
def get_zones():
    resp = requests.get(f"{BOOKING_SERVICE_URL}/zones")
    return Response(content=resp.content, status_code=resp.status_code, media_type=resp.headers.get('content-type',"application/json"))

@router.get("/zones/{zone_id}/places")
def get_places_in_zone(zone_id: int):
    resp = requests.get(f"{BOOKING_SERVICE_URL}/zones/{zone_id}/places")
    return Response(content=resp.content, status_code=resp.status_code, media_type=resp.headers.get('content-type',"application/json"))

@router.get("/places/{place_id}/slots")
def get_slots(place_id: int, request: Request):
    # Forward query parameters
    resp = requests.get(f"{BOOKING_SERVICE_URL}/places/{place_id}/slots", params=request.query_params)
    return Response(content=resp.content, status_code=resp.status_code, media_type=resp.headers.get('content-type',"application/json"))

@router.post("/")
async def create_booking(request: Request, user=Depends(get_current_user)):
    body = await request.json()
    # Forward user_id and role as headers to booking service
    headers = {
        "X-User-Id": str(user.get('user_id', user.get('sub'))),
        "X-User-Role": user.get('role', 'user')
    }
    resp = requests.post(f"{BOOKING_SERVICE_URL}/bookings", json=body, headers=headers)
    return Response(content=resp.content, status_code=resp.status_code, media_type=resp.headers.get('content-type',"application/json"))

@router.post("/by-time")
async def create_booking_by_time(request: Request, user=Depends(get_current_user)):
    body = await request.json()
    headers = {
        "X-User-Id": str(user.get('user_id', user.get('sub'))),
        "X-User-Role": user.get('role', 'user')
    }
    resp = requests.post(f"{BOOKING_SERVICE_URL}/bookings/by-time", json=body, headers=headers)
    return Response(content=resp.content, status_code=resp.status_code, media_type=resp.headers.get('content-type',"application/json"))

@router.post("/cancel")
async def cancel(request: Request, user=Depends(get_current_user)):
    body = await request.json()
    headers = {
        "X-User-Id": str(user.get('user_id', user.get('sub'))),
        "X-User-Role": user.get('role', 'user')
    }
    resp = requests.post(f"{BOOKING_SERVICE_URL}/bookings/cancel", json=body, headers=headers)
    return Response(content=resp.content, status_code=resp.status_code, media_type=resp.headers.get('content-type',"application/json"))

@router.get("/history")
async def booking_history(request: Request, user=Depends(get_current_user)):
    headers = {
        "X-User-Id": str(user.get('user_id', user.get('sub'))),
        "X-User-Role": user.get('role', 'user')
    }
    resp = requests.get(f"{BOOKING_SERVICE_URL}/bookings/history", params=request.query_params, headers=headers)
    return Response(content=resp.content, status_code=resp.status_code, media_type=resp.headers.get('content-type',"application/json"))

@router.post("/{booking_id}/extend")
async def extend_booking(booking_id: int, request: Request, user=Depends(get_current_user)):
    # // Получаем тело запроса с параметрами продления (extend_hours, extend_minutes)
    body = await request.json()
    headers = {
        "X-User-Id": str(user.get('user_id', user.get('sub'))),
        "X-User-Role": user.get('role', 'user')
    }
    # // Передаём тело запроса в booking service для корректной валидации
    resp = requests.post(f"{BOOKING_SERVICE_URL}/bookings/{booking_id}/extend", json=body, headers=headers)
    return Response(content=resp.content, status_code=resp.status_code, media_type=resp.headers.get('content-type',"application/json"))